File: Python/3kyu/the-builder-of-things.py
"""
Title: The builder of things
Link: https://www.codewars.com/kata/5571d9fc11526780a000011a
Difficulty: 3 kyu

Sentence-like DSL: is_a / is_not_a, is_the / being_the / and_the, has / having / with,
each, can (with optional archives). See the kata for full examples.
"""

from __future__ import annotations

from collections.abc import Callable, Iterator, Sequence


class Thing:
    def __init__(self, name: str) -> None:
        self.name = name

    def __getattr__(self, name: str) -> object:
        if name == "is_a":
            return _BoolChain(self, True)
        if name == "is_not_a":
            return _BoolChain(self, False)
        if name in ("is_the", "being_the", "and_the"):
            return _BeingThe(self)
        if name in ("has", "having", "with"):
            return lambda n: _Has(self, n)
        if name == "can":
            return _Can(self)
        raise AttributeError(name)


def _tag_is(thing: Thing, predicate: str, value: bool = True) -> None:
    object.__setattr__(thing, f"is_{predicate}", value)


def _singular_from_attr(attr: str, count: int) -> str:
    if count > 1 and attr.endswith("s"):
        return attr[:-1]
    return attr


class ThingSequence(Thing, Sequence):
    def __init__(self, name: str, items: list[Thing]) -> None:
        super().__init__(name)
        object.__setattr__(self, "_items", items)

    def __len__(self) -> int:
        return len(self._items)

    def __getitem__(self, index: int | slice) -> Thing | list[Thing]:
        return self._items[index]

    def __iter__(self) -> Iterator[Thing]:
        return iter(self._items)

    def each(self, fn: Callable[[Thing], None]) -> ThingSequence:
        for item in self._items:
            fn(item)
        return self


class _BoolChain:
    def __init__(self, thing: Thing, truth: bool) -> None:
        self._thing = thing
        self._truth = truth

    def __getattr__(self, noun: str) -> Thing:
        object.__setattr__(self._thing, f"is_a_{noun}", self._truth)
        object.__setattr__(self._thing, f"is_not_a_{noun}", not self._truth)
        return self._thing


class _BeingThe:
    def __init__(self, thing: Thing) -> None:
        self._thing = thing

    def __getattr__(self, prop_name: str) -> _BeingTheValue:
        return _BeingTheValue(self._thing, prop_name)


class _BeingTheValue:
    def __init__(self, thing: Thing, prop_name: str) -> None:
        self._thing = thing
        self._prop_name = prop_name

    def __getattr__(self, value_name: str) -> Thing:
        object.__setattr__(self._thing, self._prop_name, value_name)
        return self._thing


class _Has:
    def __init__(self, thing: Thing, count: int) -> None:
        self._thing = thing
        self._count = count

    def __getattr__(self, attr: str) -> Thing | ThingSequence:
        n = self._count
        base = _singular_from_attr(attr, n)

        if n > 1:
            items = []
            for _ in range(n):
                child = Thing(base)
                _tag_is(child, base)
                items.append(child)
            seq = ThingSequence(attr, items)
            _tag_is(seq, attr)
            object.__setattr__(self._thing, attr, seq)
            return seq

        child = Thing(base)
        _tag_is(child, base)
        object.__setattr__(self._thing, attr, child)
        return child


class _Can:
    def __init__(self, thing: Thing) -> None:
        self._thing = thing

    def __getattr__(self, verb: str) -> Callable[..., Thing]:
        def register(*args: object) -> Thing:
            callables = [a for a in args if callable(a)]
            if len(callables) != 1:
                raise TypeError("can.<verb> expects exactly one callable in its arguments")
            fn = callables[0]
            archives = [a for a in args if isinstance(a, str)]
            history: list[object] | None = [] if archives else None

            def bound(*call_args: object, **kwargs: object) -> object:
                out = fn(self._thing, *call_args, **kwargs)
                if history is not None:
                    history.append(out)
                return out

            object.__setattr__(self._thing, verb, bound)
            if history is not None:
                for archive_name in archives:
                    object.__setattr__(self._thing, archive_name, history)

            return self._thing

        return register

File: languages/javascript/3kyu/the-builder-of-things.js
/**
 * Title: The builder of things
 * Link: https://www.codewars.com/kata/5571d9fc11526780a000011a
 * Difficulty: 3 kyu
 *
 * Sentence-like DSL (is_a, is_the, has/having, each, can, …). See Codewars for examples.
 */

const BARE = Symbol('thingBare');

const EACH_GLOBAL_KEYS = ['having', 'being_the', 'and_the'];

function bareOf(x) {
  return x && x[BARE] ? x[BARE] : x;
}

function defGet(obj, key, get, enumerable = false) {
  Object.defineProperty(obj, key, { get, enumerable, configurable: true });
}

function emptyProxy(handlers) {
  return new Proxy({}, handlers);
}

function singularFromAttr(attr, count) {
  return count > 1 && attr.endsWith('s') ? attr.slice(0, -1) : attr;
}

function invokeCanHandler(self, fn, callArgs) {
  const g = typeof globalThis !== 'undefined' ? globalThis : {};
  const prev = g.name;
  g.name = self.name;
  try {
    return fn.apply(self, callArgs);
  } finally {
    g.name = prev;
  }
}

function eachOnArray(arr) {
  return (callback) => {
    const g = typeof globalThis !== 'undefined' ? globalThis : global;
    const saved = Object.fromEntries(EACH_GLOBAL_KEYS.map((k) => [k, g[k]]));
    try {
      for (const item of arr) {
        g.having = (n) => item.has(n);
        g.being_the = item.being_the;
        g.and_the = item.and_the;
        callback(item);
      }
    } finally {
      EACH_GLOBAL_KEYS.forEach((k) => {
        g[k] = saved[k];
      });
    }
  };
}

function canProxy(rawSelf) {
  const self = bareOf(rawSelf);
  return emptyProxy({
    get(_, verb) {
      return (...args) => {
        const last = args[args.length - 1];
        const fn = typeof last === 'function' ? last : null;
        const archives = fn ? args.slice(0, -1) : [];
        const history = archives.length ? [] : null;

        archives.forEach((archiveName) => {
          defGet(self, archiveName, () => history, true);
        });

        const method = function (...callArgs) {
          if (!fn) return undefined;
          const out = invokeCanHandler(self, fn, callArgs);
          if (history) history.push(out);
          return out;
        };

        Object.defineProperty(self, verb, {
          value: method,
          writable: true,
          enumerable: true,
          configurable: true,
        });
      };
    },
  });
}

function Thing(name) {
  const bare = Object.create(Thing.prototype);
  bare.name = name;
  return createThingProxy(bare);
}

function createThingProxy(self) {
  self[BARE] = self;

  const booleanMode = (truth) =>
    emptyProxy({
      get(_, noun) {
        defGet(self, `is_a_${noun}`, () => truth);
        defGet(self, `is_not_a_${noun}`, () => !truth);
        return createThingProxy(self);
      },
    });

  const propertyWithValue = (rawTarget = self) => {
    const target = bareOf(rawTarget);
    return emptyProxy({
      get(cache, propName) {
        if (!Object.prototype.hasOwnProperty.call(cache, propName)) {
          cache[propName] = emptyProxy({
            get(_, valueName) {
              defGet(target, propName, () => valueName, true);
              return emptyProxy({
                get(__, next) {
                  if (next === 'and_the') return propertyWithValue(target);
                  return createThingProxy(target)[next];
                },
              });
            },
          });
        }
        return cache[propName];
      },
    });
  };

  const makeHas = (rawOwner) => {
    const owner = bareOf(rawOwner);
    return (count) =>
      emptyProxy({
        get(_, attr) {
          const childBase = singularFromAttr(attr, count);

          const makeChild = () => {
            const child = new Thing(childBase);
            defGet(child, 'with', () => child.having, true);
            defGet(child, 'being_the', () => propertyWithValue(child), true);
            defGet(child, 'is_the', () => propertyWithValue(child), true);
            defGet(child, 'and_the', () => propertyWithValue(child), true);
            defGet(child, 'having', () => makeHas(child), true);
            defGet(child, 'has', () => makeHas(child), true);
            defGet(child, 'can', () => canProxy(child), true);
            return child;
          };

          if (count > 1) {
            const items = Array.from({ length: count }, makeChild);
            defGet(items, 'each', () => eachOnArray(items), false);
            owner[attr] = items;
            return items;
          }

          const one = makeChild();
          owner[attr] = one;
          return one;
        },
      });
  };

  defGet(self, 'is_a', () => booleanMode(true));
  defGet(self, 'is_not_a', () => booleanMode(false));
  defGet(self, 'is_the', () => propertyWithValue(self));
  defGet(self, 'being_the', () => propertyWithValue(self));
  defGet(self, 'and_the', () => propertyWithValue(self));
  defGet(self, 'has', () => makeHas(self));
  defGet(self, 'having', () => makeHas(self));
  defGet(self, 'can', () => canProxy(self));

  const proxy = new Proxy(self, {
    get(target, prop, receiver) {
      if (prop === 'name') return target.name;
      return Reflect.get(target, prop, receiver);
    },
    set(target, prop, value, receiver) {
      return Reflect.set(target, prop, value, receiver);
    },
  });
  proxy[BARE] = self;
  return proxy;
}

module.exports = Thing;
